File: backend/KineticEnterprise.Api/Data/LandedCost.cs
namespace KineticEnterprise.Api.Data;

/// <summary>
/// توزيع مصاريف الشحنة على أصنافها — **بالقيمة**.
///
/// <para><b>سبب وجوده:</b> صنفٌ بعشرة من المورّد وشحنةٌ بمئة وخمسين ليست
/// تكلفته عشرة. وقبل هذا كان سعر المورّد وحده هو تكلفة الصنف وتكلفة البضاعة
/// المباعة معاً — فالربح يظهر أعلى ممّا هو بمقدار الشحن كلّه، وقد تُباع
/// البضاعة بخسارةٍ والتقرير يقول ربحاً.</para>
///
/// <para><b>ولماذا بالقيمة لا بالكمية:</b> الشحن يُدفع على شحنةٍ فيها ذهبٌ
/// وأكياس. والتوزيع بالكمية يُحمّل الكيس مثل ما يُحمّل الذهب، فيخرج الكيس
/// أغلى من ثمنه ويخرج الذهب بلا شيء. والقيمة أقرب ما يكون إلى ما دُفع
/// فعلاً حين يكون الشحن على القيمة (وهو الغالب في الاستيراد: التأمين
/// والجمارك بالقيمة أصلاً).</para>
///
/// <para><b>والنصيب يُحسب على الكميّة المطلوبة لا المستلَمة:</b> الشحنة قد
/// تصل على دفعات، ونصيبُ الوحدة يجب أن يكون واحداً في الدفعة الأولى
/// والأخيرة — وإلا اختلفت تكلفة الوحدة من نفس الأمر باختلاف يوم وصولها،
/// وهو فرقٌ لا يفسّره شيء في الواقع.</para>
/// </summary>
public static class LandedCost
{
    /// <summary>سطرٌ كما يدخل الحساب: كميّته وسعر المورّد لوحدته.</summary>
    public readonly record struct Line(Guid Id, decimal Quantity, decimal UnitCost)
    {
        public decimal Value => Quantity * UnitCost;
    }

    /// <summary>
    /// نصيب الوحدة من المصاريف لكل سطر.
    ///
    /// <para>القسمة على القيمة، ثم على الكميّة — فيخرج ما يُضاف إلى سعر
    /// المورّد للوحدة الواحدة. والسطر بقيمةٍ صفر (هديّة، أو خطأ إدخال) لا
    /// يأخذ شيئاً: توزيعُ الشحن على ما لا قيمة له يُنتج قسمةً على صفر أو
    /// تحميلاً بلا أساس.</para>
    ///
    /// <para><b>والكسور تُجمَع ولا تُهمَل:</b> ثلاثة أسطر ومصاريف عشرة
    /// دنانير تُنتج ٣٫٣٣ ثلاث مرّات = ٩٫٩٩، فيضيع قرشٌ من قيمة المخزون في
    /// كل شحنة. فيُعطى الفرق لأكبر السطور قيمةً — أقلّها تأثّراً بالقرش في
    /// تكلفة وحدته.</para>
    /// </summary>
    public static Dictionary<Guid, decimal> PerUnitShares(
        IReadOnlyList<Line> lines, decimal totalCharges, int moneyDecimals = 2)
    {
        var shares = lines.ToDictionary(l => l.Id, _ => 0m);
        if (totalCharges <= 0 || lines.Count == 0) return shares;

        var totalValue = lines.Sum(l => l.Value);
        if (totalValue <= 0) return shares;

        // المبلغ المُسنَد لكل سطر أوّلاً (لا نصيب الوحدة): القرش يُجبَر على
        // مستوى السطر، وقسمتُه على الكميّة بعد ذلك تُبقي المجموع مضبوطاً.
        var assigned = new Dictionary<Guid, decimal>();
        var running = 0m;

        foreach (var line in lines)
        {
            var amount = Math.Round(totalCharges * (line.Value / totalValue), moneyDecimals,
                MidpointRounding.AwayFromZero);
            assigned[line.Id] = amount;
            running += amount;
        }

        var remainder = totalCharges - running;
        if (remainder != 0)
        {
            var biggest = lines.OrderByDescending(l => l.Value).First();
            assigned[biggest.Id] += remainder;
        }

        foreach (var line in lines)
        {
            shares[line.Id] = line.Quantity > 0 ? assigned[line.Id] / line.Quantity : 0m;
        }

        return shares;
    }

    /// <summary>
    /// التكلفة المحمَّلة للوحدة: سعر المورّد زائد نصيبها من المصاريف.
    ///
    /// <para>وهي التي تُخزَّن على الدفعة وتُحسب منها تكلفة البضاعة المباعة
    /// وتُقاس عليها الأسعار — لا سعر المورّد.</para>
    /// </summary>
    public static decimal UnitCostOf(LandedCost.Line line, Dictionary<Guid, decimal> shares) =>
        line.UnitCost + shares.GetValueOrDefault(line.Id);
}

File: lib/core/purchasing/landed_cost.dart
/// توزيع مصاريف الشحنة على أصنافها — **بالقيمة**.
///
/// <para><b>سبب وجوده في الواجهة أيضاً:</b> الحساب الرسمي في الخادم
/// (`Data/LandedCost.cs`) وهو مصدر الحقيقة. لكن التسعير يقع **قبل الحفظ**:
/// من يكتب مصاريف الشحنة ثم يضغط «سعّر بهامش ٢٥٪» يجب أن يرى التكلفة
/// المحمَّلة في اللحظة، لا بعد رحلةٍ إلى الخادم وعودة. وهامشٌ يُحسب على سعر
/// المورّد وحده خسارةٌ مقنّعة — وهو بالضبط ما تمنعه هذه الشاشة.</para>
///
/// <para>والصيغة نفسها في الملفّين، ويحرسهما اختبارٌ بالأرقام ذاتها
/// (`test/landed_cost_test.dart`) — فانحرافُ أحدهما يسقط الفحص.</para>
class LandedCost {
  const LandedCost._();

  /// نصيب الوحدة من المصاريف لكل سطر، بمفتاح ترتيب السطر في القائمة.
  ///
  /// <para>القسمة بالقيمة ثم على الكميّة. والكسور تُجمَع ولا تُهمَل: ثلاثة
  /// أسطر ومصاريف عشرة تُنتج ٣٫٣٣ ثلاث مرّات = ٩٫٩٩، فيضيع قرشٌ من قيمة
  /// المخزون في كل شحنة — فيُعطى الفرق لأكبر السطور قيمةً، أقلِّها تأثّراً
  /// بالقرش في تكلفة وحدته.</para>
  static List<double> perUnitShares({
    required List<({double quantity, double unitCost})> lines,
    required double totalCharges,
  }) {
    final shares = List<double>.filled(lines.length, 0);
    if (totalCharges <= 0 || lines.isEmpty) return shares;

    final values = lines.map((l) => l.quantity * l.unitCost).toList();
    final totalValue = values.fold<double>(0, (a, b) => a + b);
    if (totalValue <= 0) return shares;

    final assigned = List<double>.filled(lines.length, 0);
    var running = 0.0;
    for (var i = 0; i < lines.length; i++) {
      assigned[i] = _round2(totalCharges * (values[i] / totalValue));
      running += assigned[i];
    }

    final remainder = _round2(totalCharges - running);
    if (remainder != 0) {
      var biggest = 0;
      for (var i = 1; i < values.length; i++) {
        if (values[i] > values[biggest]) biggest = i;
      }
      assigned[biggest] = _round2(assigned[biggest] + remainder);
    }

    for (var i = 0; i < lines.length; i++) {
      shares[i] = lines[i].quantity > 0 ? assigned[i] / lines[i].quantity : 0;
    }
    return shares;
  }

  /// تقريبٌ إلى قرشين بعيداً عن الصفر — مطابقٌ لـ`MidpointRounding.AwayFromZero`
  /// في الخادم. وتقريبُ دارت الافتراضي إلى الزوجي يُنتج فرقاً بالقرش بين
  /// ما تراه الشاشة وما يحفظه الخادم.
  static double _round2(double value) {
    final scaled = value * 100;
    final rounded = scaled < 0 ? -(scaled.abs() + 0.5).floorToDouble() : (scaled + 0.5).floorToDouble();
    return rounded / 100;
  }
}
